### ISSUE TITLE
Inaccurate recognition of lifetime

### ISSUE THREAD

**Issue by xizheyin** [rust-lang/rust#125878]
I tried this code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=7c66cdf7dd02f2fbc8cd2f89fe91114b
It report error, but when I switch test1() and test2() around, no errors are reported.
Shouldn't both cases be the same, both reporting errors or neither.

```rust
#[derive(Debug)]
struct A<'t> {
    value: &'t i32,
}

impl A<'_> {
    fn test1(& self) {}

    fn test2(&mut self) {}

    fn test3<'guess>(&'guess mut self) {}
}

fn main() {
    let v = 1;
    let mut a = A { value: &v };
    a.test2();
    a.test1();
}

```

## Error report
```
   Compiling playground v0.0.1 (/playground)
error[E0502]: cannot borrow `a` as immutable because it is also borrowed as mutable
  --> src/main.rs:15:5
   |
14 |     a.test2();
   |     - mutable borrow occurs here
15 |     a.test1();
   |     ^
   |     |
   |     immutable borrow occurs here
   |     mutable borrow later used here

For more information about this error, try `rustc --explain E0502`.
error: could not compile `playground` (bin "playground") due to 1 previous error
```
</p>
</details>

**Comment by theemathas** [comment]
Not a bug.

The code in the issue description does not produce any errors. However, the code in the playground link is different:

```rust
#[derive(Debug)]
struct A<'t> {
    value: &'t i32,
}

impl<'source> A<'source> {
    fn test1(&'source self) {}
    fn test2(&'source mut self) {}
}

fn main() {
    let v = 1;
    let mut a = A { value: &v };
    a.test2();
    a.test1();
    
}
```

This code produces the error in the issue description. This is because the test2() method takes an argument of type `&'source mut A<'source>`. This function signature potentially allows the test2 function to replace the value of `a` with a struct `A` that contains a mutable reference to itself, therefore preventing creating further references to `a`.

**Comment by xizheyin** [comment]
ok, Thanks!
